### PR TITLE
Memoize parsing and symbol detection: improves performance by removing repeated I/O and parsing

### PR DESCRIPTION
**Pull request by Ocramius**
Requires #337 to be finished first.

### CODE REVIEW

File: test/unit/SingletonConfiguration.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * a new instance of a shared one.
+ */
+final class SingletonConfiguration
+{
+    /**
+     * @var Configuration|null
+     */
+    private static $configuration;
+
+    public static function configuration() : Configuration
```
Comment (Ocramius) [inline test/unit/SingletonConfiguration.php]: This makes the test suite ~40% faster, so I think it's worth keeping it for now.

File: src/SourceLocator/Type/MemoizingSourceLocator.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+declare(strict_types=1);
+
+namespace Roave\BetterReflection\SourceLocator\Type;
+
+use Roave\BetterReflection\Identifier\Identifier;
+use Roave\BetterReflection\Identifier\IdentifierType;
+use Roave\BetterReflection\Reflection\Reflection;
+use Roave\BetterReflection\Reflector\Reflector;
+
+final class MemoizingSourceLocator implements SourceLocator
```
Comment (Ocramius) [inline src/SourceLocator/Type/MemoizingSourceLocator.php]: TODO: test this.

Also, this makes no huge difference on the test suite itself (the suite manually instantiates locators most of the time)

File: src/SourceLocator/Ast/Parser/MemoizingParser.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+namespace Roave\BetterReflection\SourceLocator\Ast\Parser;
+
+use PhpParser\ErrorHandler;
+use PhpParser\Node;
+use PhpParser\Parser;
+
+/**
+ * @internal
+ */
+final class MemoizingParser implements Parser
```
Comment (Ocramius) [inline src/SourceLocator/Ast/Parser/MemoizingParser.php]: TODO: test this

Comment (Ocramius) [reply src/SourceLocator/Ast/Parser/MemoizingParser.php]: Tested

File: src/SourceLocator/Type/MemoizingSourceLocator.php (hunk 2 above)
Comment (Ocramius) [reply src/SourceLocator/Type/MemoizingSourceLocator.php]: Tested

**Comment by Ocramius** [comment]
@asgrim this is ready to be :ship: -ed

File: src/SourceLocator/Ast/Parser/MemoizingParser.php (hunk 3 above)
Comment (TomasVotruba) [reply src/SourceLocator/Ast/Parser/MemoizingParser.php]: Is that intentional name or typo in `Memorizing`?

Comment (Ocramius) [reply src/SourceLocator/Ast/Parser/MemoizingParser.php]: https://en.m.wikipedia.org/wiki/Memoization

**Comment by kukulich** [comment]
It works very well!

**Comment by TomasVotruba** [comment]
:+1:

File: src/SourceLocator/Ast/Parser/MemoizingParser.php (hunk 3 above)
Comment (TomasVotruba) [reply src/SourceLocator/Ast/Parser/MemoizingParser.php]: Oh, thanks for new term

**Comment by Ocramius** [comment]
I see :green_apple:, therefore I :ship:

**Comment by asgrim** [comment]
BEAT ME TO IT haha

**Comment by TomasVotruba** [comment]
:comet: :tada:

**Pull request merged**